teuthology/task/kube_spray.py

```diff
@@ -0,0 +1,196 @@
+from teuthology.orchestra import run
+from teuthology.misc import install_package
+log = logging.getLogger(__name__)
+
+
+class kube_setup(Task):
+    name = 'kube_setup'
+
+    """
+    A task to setup kubernetes cluster using kubespray
+
+    - kube_setup:
+        repo: {git_base}kube-spray.git
+        release: v2.8.2 (default is v2.8.3)
+        vars:
+            cert_management: None
+            var1: value1
+
+        Roles Usage:
+        [Kmaster.0, Ketcd.0]
+        [mon.a, Knode.0]
+        [osd.0, Knode.1]
+
+        Task Usage:
+          - ssh-keys:
+          - kube_setup:
+    """
+
+    groups_to_roles = dict(
+            all='K',
+            master='Kmaster',
+            etcd='Ketcd',
+            node='Knode',
+        )
+    kubespray_git = 'https://github.com/kubernetes-sigs/kubespray.git'
+
+    def __init__(self, ctx, config):
+        super(kube_setup, self).__init__(ctx, config)
+        config = self.config or dict()
+        self.playbook = None
+        self.release = config.get('release', 'v2.8.3')
+        if 'repo' not in config:
+            self.config['repo'] = self.kubespray_git
+        if 'vars' not in config:
+            vars = dict()
+            config['vars'] = vars
+        vars = config['vars']
+
+    def setup(self):
+        super(kube_setup, self).setup()
+        # generate hosts file based on test config
+        self.generate_hosts_file()
+        # everything from vars in config go into group_vars/all file
+        extra_vars = dict()
+        extra_vars.update(self.config.get('vars', dict()))
+        self.extra_vars_file = None
+        if len(extra_vars) > 0:
+            gvar = yaml.dump(extra_vars, default_flow_style=False)
+            self.extra_vars_file = self._write_hosts_file(
+                                        prefix='kube_spray_gvar',
+                                        content=gvar)
+
+    def begin(self):
+        super(kube_setup, self).begin()
+        self.execute_playbook()
+
+    def generate_hosts_file(self):
+        hosts_dict = dict()
+        for group in sorted(self.groups_to_roles.keys()):
+            role_prefix = self.groups_to_roles[group]
+            want = lambda role: role.startswith(role_prefix)
+            for (remote, roles) in self.cluster.only(want).remotes.iteritems():
+                hostname = remote.hostname
+                if group.startswith('master') or group.startswith('node'):
+                    group_name = 'kube-' + group
+                else:
+                    group_name = group
+                if group_name not in hosts_dict:
+                    hosts_dict[group_name] = {}
+
+                if group_name == 'all':
+                    vars = "ip={ip} ansible_host={ip}".format(ip=remote.ip_address)
+                    hosts_dict[group_name][hostname] = hostname + " " + vars
+                else:
+                    hosts_dict[group_name][hostname] = hostname
+
+        hosts_stringio = StringIO()
```

Comment:
Any reasons to use StringIO instead basic strings here?

Reply:
lt provides file like operations, you can seek(0) and dump the buffer to a file as done later in the code.

Reply:
using seek() operations and stream for simple string is overkill for python